source/logrepl/snapshot.go

```diff
@@ -0,0 +1,207 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logrepl
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/conduitio/conduit-connector-postgres/pgutil"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+)
+
+var psql = sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
+
+const actionSnapshot string = "snapshot"
+
+type SnapshotConfig struct {
+	URI          string
+	SnapshotName string
+	Table        string
+	Columns      []string
+	KeyColumn    string
+}
+
+type SnapshotIterator struct {
+	config SnapshotConfig
+
+	conn *pgx.Conn
+	rows pgx.Rows
+
+	internalPos int64
+}
+
+func NewSnapshotIterator(ctx context.Context, conn *pgx.Conn, cfg SnapshotConfig) (*SnapshotIterator, error) {
+	s := &SnapshotIterator{
+		config: cfg,
+		conn:   conn,
+	}
+
+	err := s.startSnapshotTx(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start snapshot tx: %w", err)
+	}
+
+	err = s.loadRows(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load rows: %w", err)
+	}
+
+	return s, nil
+}
+
+func (s *SnapshotIterator) loadRows(ctx context.Context) error {
+	query, args, err := psql.
+		Select(s.config.Columns...).
+		From(s.config.Table).
+		ToSql()
+	if err != nil {
+		return fmt.Errorf("failed to create read query: %w", err)
+	}
+
+	rows, err := s.conn.Query(ctx, query, args...)
+	if err != nil {
+		return fmt.Errorf("failed to query rows: %w", err)
+	}
+	s.rows = rows
+
+	return nil
+}
+
+func (s *SnapshotIterator) startSnapshotTx(ctx context.Context) error {
+	_, err := s.conn.Exec(ctx, `BEGIN ISOLATION LEVEL REPEATABLE READ;`)
+	if err != nil {
+		return nil
+	}
+
+	snapshotTx := fmt.Sprintf(`SET TRANSACTION SNAPSHOT '%s'`, s.config.SnapshotName)
+	_, err = s.conn.Exec(ctx, snapshotTx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *SnapshotIterator) Next(ctx context.Context) (sdk.Record, error) {
+	if !s.rows.Next() {
+		if err := s.rows.Err(); err != nil {
+			return sdk.Record{}, fmt.Errorf("rows error: %w", err)
+		}
+
+		return sdk.Record{}, sdk.ErrBackoffRetry
```

Comment:
`ErrBackoffRetry` doesn't seem like the right error here, if there are no more records there's no reason to retry. We should probably create an error to signal that the snapshot reached the end, this way the combined iterator can check for the error and switch to the CDC iterator.

Reply:
Addressed in #27.